src/rulesets/Base/array.jl

```diff
@@ -24,66 +24,175 @@ end
 ##### `hcat` (🐈)
 #####
 
-function rrule(::typeof(hcat), A::AbstractArray, Bs::AbstractArray...)
-    function hcat_pullback(Ȳ)
-        Xs = (A, Bs...)
-        ntuple(length(Bs) + 2) do full_i
-            full_i == 1 && return NoTangent()
-
-            i = full_i - 1
-            l = mapreduce(j->size(Xs[j], 2), Base.add_sum, 1:i-1; init=0)
-            u = l + size(Xs[i], 2)
-            dim = u > l + 1 ? (l+1:u) : u
-            # NOTE: The copy here is defensive, since `selectdim` returns a view which we can
-            # materialize with `copy`
-            copy(selectdim(Ȳ, 2, dim))
+using Compat # get here needs Compat 3.31
+
+function rrule(::typeof(hcat), Xs...)
+    Y = hcat(Xs...)  # note that Y always has 1-based indexing, even if X isa OffsetArray
+    ndimsY = Val(ndims(Y))  # this avoids closing over Y, Val() is essential for type-stability
+    sizes = map(size, Xs)   # this avoids closing over Xs
+    function 🐈_pullback(dY)
+        hi = Ref(0)  # Ref avoids hi::Core.Box
+        dXs = map(sizes) do sizeX
+            ndimsX = length(sizeX)
+            lo = hi[] + 1
+            hi[] += get(sizeX, 2, 1)
+            ind = ntuple(ndimsY) do d
+                if d==2
+                    d > ndimsX ? lo : lo:hi[]
+                else
+                    d > ndimsX ? 1 : (:)
+                end
+            end
+            if ndimsX > 0
+                InplaceableThunk(@thunk(dY[ind...]), dX -> dX .+= view(dY, ind...))
+            else
+                dY[ind...]
+            end
```

Comment:
I guess for `vcat(1,2,3)` this frees `dY`. 

Not sure whether `@thunk` is close enough to zero cost for scalar arguments? Haven't tried to time that.

Also, for `vcat(1, cu([2,3]))` this ought to be `getindex_allowing_scalar(dY, ind...)`. Haven't thought more about where that ought to live.

Reply:
Also, any idea why this should fail to infer? As here https://github.com/JuliaDiff/ChainRules.jl/pull/451#discussion_r654816686, it did work before the `InplaceableThunk(@thunk(...`.

Reply:
CUDA attempts:
```julia
julia> v = cu(rand(3));

julia> CUDA.@allowscalar v[2]  # the official way
0.08611474f0

julia> @macroexpand CUDA.@allowscalar v[2]
quote
    GPUArrays.task_local_storage(:ScalarIndexing, GPUArrays.ScalarAllowed) do 
        v[2]
    end
end

julia> sum(view(v,2:2))  # another idea, not easy in N dims?
0.08611474f0

julia> sum(view(v,2))  # another idea
ERROR: MethodError: no method matching ndims(::Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{0}, Nothing, typeof(identity), Tuple{CuArray{Float32, 0}}})
Closest candidates are:
...
```

Reply:
I know nothing about GPU, we can probably poke vchuravy, if we need to for help with that.

Re: inference:
It might be worth just not using Thunks (or InplaceThunks) for now, 
and we can add them back in later

Reply:
I made https://github.com/JuliaGPU/GPUArrays.jl/pull/363 after asking around. Maybe Base should also specialise...

Reply:
Inference now only fails on 1.0. Somehow I thought that got auto-skipped?

Reply:
I think it gets auto-skipped on a case by case basis maybe?

Reply:
It only gets skipped for PkgEval, our CI always runs these. I agree that it might make sense to always disable them on 1.0 though.

Reply:
OK, thanks. Disabled by hand now, for 1.0, hope I got the right ones!